codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,136 +16,248 @@
+ * http://netty.io/wiki/reference-counted-objects.html
+ *
+ * <h3>HTTP Upgrade</h3>
+ *
+ * Server-side HTTP to HTTP/2 upgrade is supported in conjunction with {@link Http2ServerUpgradeCodec}; the necessary
+ * HTTP-to-HTTP/2 conversion is performed automatically.
  */
 @UnstableApi
 public class Http2FrameCodec extends ChannelDuplexHandler {
 
-    private static final Http2FrameLogger HTTP2_FRAME_LOGGER = new Http2FrameLogger(INFO, Http2FrameCodec.class);
+    private static final InternalLogger LOG = InternalLoggerFactory.getInstance(Http2FrameCodec.class);
 
     private final Http2ConnectionHandler http2Handler;
     private final boolean server;
+    private final PropertyKey streamKey;
+
+    // Used to adjust flow control window on channel active. Set to null afterwards.
+    private Integer initialLocalConnectionWindow;
 
     private ChannelHandlerContext ctx;
     private ChannelHandlerContext http2HandlerCtx;
 
+    private Http2Stream2Impl pendingOutboundStreamsTail;
+
+    /** Lock protecting modifications to idle outbound streams. **/
+    private final Object lock = new Object();
+
+    /** Number of buffered streams if the {@link StreamBufferingEncoder} is used. **/
+    private int numBufferedStreams;
+
     /**
-     * Construct a new handler.
-     *
-     * @param server {@code true} this is a server
+     * Create a new handler. Use {@link Http2FrameCodecBuilder}.
      */
-    public Http2FrameCodec(boolean server) {
-        this(server, HTTP2_FRAME_LOGGER);
+    Http2FrameCodec(Http2ConnectionEncoder encoder, Http2ConnectionDecoder decoder, Http2Settings initialSettings,
+                    long gracefulShutdownTimeoutMillis) {
+        decoder.frameListener(new FrameListener());
+        http2Handler = new InternalHttp2ConnectionHandler(decoder, encoder, initialSettings);
+        http2Handler.connection().addListener(new ConnectionListener());
+        http2Handler.gracefulShutdownTimeoutMillis(gracefulShutdownTimeoutMillis);
+        server = http2Handler.connection().isServer();
+        streamKey = connection().newKey();
+        initialLocalConnectionWindow = initialSettings.initialWindowSize();
+    }
+
+    Http2ConnectionHandler connectionHandler() {
+        return http2Handler;
     }
 
     /**
-     * Construct a new handler.
+     * Creates a new outbound/local stream.
+     *
+     * <p>The object is added to a list of idle streams, so that in case the stream object is never made active, the
+     * {@link Http2Stream2#closeFuture()} still completes.
+     *
+     * <p>This method may only be called after the handler has been added to a {@link io.netty.channel.ChannelPipeline}.
      *
-     * @param server {@code true} this is a server
+     * <p>This method is thread-safe.
      */
-    public Http2FrameCodec(boolean server, Http2FrameLogger frameLogger) {
-        this(server, new DefaultHttp2FrameWriter(), frameLogger);
-    }
+    // TODO(buchgr): Discuss: Should this method be thread safe?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)